ngmix/runners.py

```diff
@@ -0,0 +1,208 @@
+    """
+    def go(self, obs):
+        """
+        Run the fitter on the input observation(s), possibly multiple times
+        using guesses generated from the guesser
+
+        Parameters
+        ----------
+        obs: ngmix Observation(s)
+            Observation, ObsList, or MultiBandObsList
+
+        Returns
+        -------
+        result dictionary
+        """
+
+        run_fitter(
+            obs=obs, fitter=self.fitter, guesser=self.guesser, ntry=self.ntry,
+        )
+
+    def get_result(self):
+        """
+        get the result dict
+        """
+        return self.fitter.get_result()
+
+
+class PSFRunner(RunnerBase):
+    """
+    Run a fitter on each psf observation.
+
+    Parameters
+    ----------
+    fitter: ngmix fitter or measurer
+        An object to perform measurements, must have a go(obs=obs, guess=guess)
+        method.
+    guesser: ngmix guesser object
+        Must be a callable returning an array of parameters
+    ntry: int, optional
+        Number of times to try if there is failure
+    """
+
+    def go(self, obs, set_result=False):
+        """
+        Run the fitter on the psf observations associated with the input
+        observation(s), possibly multiple times using guesses generated from
+        the guesser.
+
+        Parameters
+        ----------
+        obs: ngmix Observation(s)
+            Observation, ObsList, or MultiBandObsList
+        ntry: int, optional
+            Number of times to try if there is failure
+        set_result: bool, optional
+            If set to True, the meta['result'] and the .gmix attribute
+
+        Side Effects
+        ------------
+        If set_result is True then .meta['result'] is set to the fit result and the
+        .gmix attribuite is set for each successful fit
+        """
+
+        run_psf_fitter(
+            obs=obs, fitter=self.fitter, guesser=self.guesser, ntry=self.ntry,
+            set_result=set_result,
+        )
+
+
+def run_fitter(obs, fitter, guesser=None, ntry=1):
+    """
+    run a fitter multiple times if needed, with guesses generated from the
+    input guesser
+
+    Parameters
+    ----------
+    obs: ngmix Observation(s)
+        Observation, ObsList, or MultiBandObsList
+    fitter: ngmix fitter or measurer
+        An object to perform measurements, must have a go(obs=obs, guess=guess)
+        method.
+    guesser: ngmix guesser object, optional
+        Must be a callable returning an array of parameters
+    ntry: int, optional
+        Number of times to try if there is failure
+
+    Returns
+    -------
+    result dictionary
+    """
+
+    for i in range(ntry):
+
+        if guesser is not None:
+            guess = guesser(obs=obs)
+            fitter.go(obs=obs, guess=guess)
+        else:
+            fitter.go(obs=obs)
+
+        res = fitter.get_result()
+        if res['flags'] == 0:
+            break
+
+
+def run_psf_fitter(obs, fitter, guesser=None, ntry=1, set_result=False):
```

Comment:
Is set_result=False the right default here? It seems like unless that is true this function doesn't do a lot. What is the use case for not setting a result?

Reply:
Originally I had made all these functions pure by default, and returning the results instead of setting them.  But I realized that was impractical for almost all use cases we care about, because we need the gmix set.

We should make this the only behavior rather than optional.